### ISSUE TITLE
Cannot install mmmagic on windows

### ISSUE THREAD

**Issue by yanakey** [mscdex/mmmagic#80]
Hello!
Cannot install mmmagic on windows:

```
$ npm install mmmagic

> mmmagic@0.4.3 install D:\projects\node_modules\mmmagic
> node-gyp rebuild


D:\projects\node_modules\mmmagic>if not defined npm_config_node_gyp (node "D:\Soft\DEV\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuil
d )  else (node "" rebuild )
Сборка проектов в этом решении по одному. Чтобы включить параллельную сборку, добавьте параметр "/m".
  apprentice.c
  apptype.c
  ascmagic.c
  cdf.c
  cdf_time.c
  compress.c
  der.c
  encoding.c
  fsmagic.c
..\..\..\deps\libmagic\src\der.c(198): warning C4013: 'snprintf' undefined; assuming extern returning int [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\der.c(275): error C2065: '__func__' : undeclared identifier [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\der.c(300): error C2065: '__func__' : undeclared identifier [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\der.c(310): error C2065: '__func__' : undeclared identifier [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagic.vcxproj]
  funcs.c
  is_tar.c
  magic.c
..\..\..\deps\libmagic\src\cdf_time.c(137): warning C4244: '=' : conversion from 'time_t' to 'long', possible loss of data [D:\projects\node_modules\mmmagic\build\deps\libmagic\li
bmagic.vcxproj]
..\..\..\deps\libmagic\src\cdf.c(1087): warning C4013: 'snprintf' undefined; assuming extern returning int [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\cdf_time.c(174): warning C4013: 'snprintf' undefined; assuming extern returning int [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagic.vcxpr
oj]
  print.c
  readcdf.c
  readelf.c
  softmagic.c
  fmtcheck.c
  asctime_r.c
  asprintf.c
..\..\..\deps\libmagic\src\readcdf.c(133): warning C4013: 'strcasestr' undefined; assuming extern returning int [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagic.vcxp
roj]
..\..\..\deps\libmagic\src\readcdf.c(133): warning C4047: '!=' : 'int' differs in levels of indirection from 'void *' [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagi
c.vcxproj]
..\..\..\deps\libmagic\src\readelf.c(861): warning C4244: 'return' : conversion from 'uint64_t' to 'off_t', possible loss of data [D:\projects\node_modules\mmmagic\build\deps\libm
agic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\readcdf.c(239): warning C4133: 'function' : incompatible types - from 'long *' to 'const time_t *' [D:\projects\node_modules\mmmagic\build\deps\libmagic
\libmagic.vcxproj]
..\..\..\deps\libmagic\src\softmagic.c(461): warning C4013: 'snprintf' undefined; assuming extern returning int [D:\projects\node_modules\mmmagic\build\deps\libmagic\libmagic.vcxp
roj]
  ctime_r.c
  dprintf.c
  getline.c
  gmtime_r.c
  localtime_r.c
  pread.c
  snprintf.c
  strcasestr.c
  strlcat.c
  strlcpy.c
  vasprintf.c
  win_delay_load_hook.c
  pcre_byte_order.c
  pcre_chartables.c
  pcre_compile.c
  pcre_config.c
  pcre_dfa_exec.c
  pcre_exec.c
  pcre_fullinfo.c
  pcre_get.c
  pcre_globals.c
  pcre_jit_compile.c
  pcre_maketables.c
..\..\..\..\deps\libmagic\pcre\pcre_dfa_exec.c(1662): warning C4146: unary minus operator applied to unsigned type, result still unsigned [D:\projects\node_modules\mmmagic\build\d
eps\libmagic\pcre\libpcre.vcxproj]
..\..\..\..\deps\libmagic\pcre\pcre_dfa_exec.c(1708): warning C4146: unary minus operator applied to unsigned type, result still unsigned [D:\projects\node_modules\mmmagic\build\d
eps\libmagic\pcre\libpcre.vcxproj]
..\..\..\..\deps\libmagic\pcre\pcre_dfa_exec.c(1752): warning C4146: unary minus operator applied to unsigned type, result still unsigned [D:\projects\node_modules\mmmagic\build\d
eps\libmagic\pcre\libpcre.vcxproj]
..\..\..\..\deps\libmagic\pcre\pcre_dfa_exec.c(1793): warning C4146: unary minus operator applied to unsigned type, result still unsigned [D:\projects\node_modules\mmmagic\build\d
eps\libmagic\pcre\libpcre.vcxproj]
  pcre_newline.c
  pcre_ord2utf8.c
  pcre_refcount.c
  pcre_string_utils.c
  pcre_study.c
  pcre_tables.c
  pcre_ucd.c
  pcre_valid_utf8.c
  pcre_version.c
  pcre_xclass.c
  pcreposix.c
  win_delay_load_hook.c
  libpcre.vcxproj -> D:\projects\node_modules\mmmagic\build\Release\\libpcre.lib
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\12.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (D:\Soft\DEV\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:204:12)
gyp ERR! System Windows_NT 6.2.9200
gyp ERR! command "d:\\Soft\\DEV\\nodejs\\node.exe" "D:\\Soft\\DEV\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd D:\projects\node_modules\mmmagic
gyp ERR! node -v v6.0.0
gyp ERR! node-gyp -v v3.3.1
gyp ERR! not ok
D:\projects
`-- node-gyp@3.3.1

npm WARN enoent ENOENT: no such file or directory, open 'D:\projects\package.json'
npm WARN projects No description
npm WARN projects No repository field.
npm WARN projects No README data
npm WARN projects No license field.
npm ERR! Windows_NT 6.2.9200
npm ERR! argv "d:\\Soft\\DEV\\nodejs\\node.exe" "d:\\Soft\\DEV\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "mmmagic"
npm ERR! node v6.0.0
npm ERR! npm  v3.8.6
npm ERR! code ELIFECYCLE

npm ERR! mmmagic@0.4.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the mmmagic@0.4.3 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the mmmagic package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs mmmagic
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls mmmagic
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     D:\projects\npm-debug.log

```

```
$ git clone https://github.com/mscdex/mmmagic.git
Cloning into 'mmmagic'...
remote: Counting objects: 1616, done.
Receiving objects: 100% (1616/1616), 4.20 MiB | 1.29 MiB/s, done.16

Resolving deltas: 100% (728/728), done.
Checking connectivity... done.

yanot@VAIO /d/projects/mmmagic (master)
$ npm i

> mmmagic@0.4.3 install d:\projects\mmmagic
> node-gyp rebuild


d:\projects\mmmagic>if not defined npm_config_node_gyp (node "D:\Soft\DEV\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (no
de "" rebuild )
Сборка проектов в этом решении по одному. Чтобы включить параллельную сборку, добавьте параметр "/m".
  apprentice.c
  apptype.c
  ascmagic.c
  cdf.c
  cdf_time.c
  compress.c
  der.c
  encoding.c
..\..\..\deps\libmagic\src\der.c(198): warning C4013: 'snprintf' undefined; assuming extern returning int [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\der.c(275): error C2065: '__func__' : undeclared identifier [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\der.c(300): error C2065: '__func__' : undeclared identifier [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\der.c(310): error C2065: '__func__' : undeclared identifier [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
  fsmagic.c
  funcs.c
..\..\..\deps\libmagic\src\cdf_time.c(137): warning C4244: '=' : conversion from 'time_t' to 'long', possible loss of data [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxpro
j]
  is_tar.c
  magic.c
..\..\..\deps\libmagic\src\cdf_time.c(174): warning C4013: 'snprintf' undefined; assuming extern returning int [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\cdf.c(1087): warning C4013: 'snprintf' undefined; assuming extern returning int [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
  print.c
  readcdf.c
  readelf.c
  softmagic.c
  fmtcheck.c
  asctime_r.c
..\..\..\deps\libmagic\src\readelf.c(861): warning C4244: 'return' : conversion from 'uint64_t' to 'off_t', possible loss of data [d:\projects\mmmagic\build\deps\libmagic\libmagic
.vcxproj]
  asprintf.c
..\..\..\deps\libmagic\src\readcdf.c(133): warning C4013: 'strcasestr' undefined; assuming extern returning int [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\readcdf.c(133): warning C4047: '!=' : 'int' differs in levels of indirection from 'void *' [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\softmagic.c(461): warning C4013: 'snprintf' undefined; assuming extern returning int [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcxproj]
..\..\..\deps\libmagic\src\readcdf.c(239): warning C4133: 'function' : incompatible types - from 'long *' to 'const time_t *' [d:\projects\mmmagic\build\deps\libmagic\libmagic.vcx
proj]
  ctime_r.c
  dprintf.c
  getline.c
  gmtime_r.c
  localtime_r.c
  pread.c
  snprintf.c
  strcasestr.c
  strlcat.c
  strlcpy.c
  vasprintf.c
  win_delay_load_hook.c
  pcre_byte_order.c
  pcre_chartables.c
  pcre_compile.c
  pcre_config.c
  pcre_dfa_exec.c
  pcre_exec.c
  pcre_fullinfo.c
  pcre_get.c
  pcre_globals.c
  pcre_jit_compile.c
  pcre_maketables.c
..\..\..\..\deps\libmagic\pcre\pcre_dfa_exec.c(1662): warning C4146: unary minus operator applied to unsigned type, result still unsigned [d:\projects\mmmagic\build\deps\libmagic\
pcre\libpcre.vcxproj]
..\..\..\..\deps\libmagic\pcre\pcre_dfa_exec.c(1708): warning C4146: unary minus operator applied to unsigned type, result still unsigned [d:\projects\mmmagic\build\deps\libmagic\
pcre\libpcre.vcxproj]
..\..\..\..\deps\libmagic\pcre\pcre_dfa_exec.c(1752): warning C4146: unary minus operator applied to unsigned type, result still unsigned [d:\projects\mmmagic\build\deps\libmagic\
pcre\libpcre.vcxproj]
..\..\..\..\deps\libmagic\pcre\pcre_dfa_exec.c(1793): warning C4146: unary minus operator applied to unsigned type, result still unsigned [d:\projects\mmmagic\build\deps\libmagic\
pcre\libpcre.vcxproj]
  pcre_newline.c
  pcre_ord2utf8.c
  pcre_refcount.c
  pcre_string_utils.c
  pcre_study.c
  pcre_tables.c
  pcre_ucd.c
  pcre_valid_utf8.c
  pcre_version.c
  pcre_xclass.c
  pcreposix.c
  win_delay_load_hook.c
  libpcre.vcxproj -> d:\projects\mmmagic\build\Release\\libpcre.lib
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\12.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (D:\Soft\DEV\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:204:12)
gyp ERR! System Windows_NT 6.2.9200
gyp ERR! command "d:\\Soft\\DEV\\nodejs\\node.exe" "D:\\Soft\\DEV\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd d:\projects\mmmagic
gyp ERR! node -v v6.0.0
gyp ERR! node-gyp -v v3.3.1
gyp ERR! not ok

npm WARN mmmagic@0.4.3 No license field.
npm ERR! Windows_NT 6.2.9200
npm ERR! argv "d:\\Soft\\DEV\\nodejs\\node.exe" "d:\\Soft\\DEV\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "i"
npm ERR! node v6.0.0
npm ERR! npm  v3.8.6
npm ERR! code ELIFECYCLE
npm ERR! mmmagic@0.4.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the mmmagic@0.4.3 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the mmmagic package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs mmmagic
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls mmmagic
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     d:\projects\mmmagic\npm-debug.log

```

```
$ node -v
v6.0.0

$ npm -v
3.8.6
```

**Comment by mscdex** [comment]
What version of Visual Studio are you using to compile this?

**Comment by yanakey** [comment]
12.0

**Comment by mscdex** [comment]
Can you try applying the patch below and let me know if that fixes things?:

```diff
diff --git a/deps/libmagic/src/der.c b/deps/libmagic/src/der.c
index e7f39bc..2cfe06f 100644
--- a/deps/libmagic/src/der.c
+++ b/deps/libmagic/src/der.c
@@ -56,6 +56,11 @@ FILE_RCSID("@(#)$File: der.c,v 1.4 2016/03/21 23:04:40 christos Exp $")
 #include <string.h>
 #include <ctype.h>
 
+// XXX: addition by mscdex
+#ifdef _MSC_VER
+#define __func__ __FUNCTION__
+#endif
+
 #ifndef TEST_DER
 #include "magic.h"
 #include "der.h"
```

**Comment by yanakey** [comment]
I've applied this change to der.c, now it is OK!

**Comment by mscdex** [comment]
Thanks for verification, fix landed in 98b6ce2ce0657a4277c0c66d63f19ed01621b4a8.
